### PR TITLE
Delay revalidation after edit and improve cancel checking

### PR DESCRIPTION
**Pull request by datho7561**
Delays revalidation of edited Qute templates to prevent excessive validation requests. Improves cancel checking in completion.

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 1]
```diff
@@ -332,84 +343,67 @@ public QuteTextDocument getDocument(String uri) {
...
-				.thenCompose(a -> {
-					return a;
-				});
-		CancelChecker cancelIndicator = () -> {
-			if (result.isCancelled())
-				throw new CancellationException();
-		};
-		start.complete(cancelIndicator);
-		return result;
+	public <R> CompletableFuture<R> getTemplateCompose(TextDocumentIdentifier documentIdentifier,
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:380]: It seems that LSP4MP has the same problem with findDefinition. Also I suggest that you create in the TextDocuments a new `computeModelAsync`  static method which does the compose with the following signature in `ModelTextDocuments`:

```java
public <R> CompletableFuture<R> computeModelAsyncCompose(TextDocumentIdentifier documentIdentifier,
			BiFunction<T, CancelChecker, CompletableFuture<R>> code) {
```

and if it works, we should change too LSPMP to use this new static method `ModelTextDocuments#computeModelAsyncCompose`

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 2]
```diff
@@ -119,38 +124,40 @@ public void didSave(DidSaveTextDocumentParams params) {
...
-		return computeModelAsync2(getDocument(params.getTextDocument().getUri()).getModel(),
-				(cancelChecker, template) -> {
+		return getTemplateCompose(params.getTextDocument(),
+				(template, cancelChecker) -> {
 					return getQuteLanguageService()
 							.doComplete(template, params.getPosition(), sharedSettings.getCompletionSettings(),
 									sharedSettings.getFormattingSettings(), sharedSettings.getNativeSettings(),
 									cancelChecker) //
-							.thenApply(list -> {
+							.thenApplyAsync(list -> {
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:133]: Why `thenApplyAsync` ?

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 3]
```diff
@@ -332,84 +343,67 @@ public QuteTextDocument getDocument(String uri) {
...
+		}).thenCompose((tuple) -> {
+			return code.apply(tuple.getFirst(), tuple.getSecond());
+		});
 	}
 
 	public void validationSettingsChanged() {
 		// trigger validation for all opened Qute template files
 		documents.all().stream().forEach(document -> {
-			triggerValidationFor((QuteTextDocument) document);
+			triggerValidationFor((QuteTextDocument) document, true);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:392]: No need to delay the validation after a settings changed.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java [hunk 4]
```diff
@@ -332,84 +343,67 @@ public QuteTextDocument getDocument(String uri) {
...
 		});
 	}
 
 	public void dataModelChanged(JavaDataModelChangeEvent event) {
 		// trigger validation for all opened Qute template files
 		documents.all().stream().forEach(document -> {
-			triggerValidationFor((QuteTextDocument) document);
+			triggerValidationFor((QuteTextDocument) document, true);
```
Comment (angelozerr) [inline qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:399]: No need to delay the validation after a data model changed.

`triggerValidationFor((QuteTextDocument) document, false);`

**Comment by angelozerr** [comment]
@datho7561 just for your information, the delay validation is very helpfull when you use linked editing range https://github.com/redhat-developer/vscode-quarkus/blob/master/docs/qute/TemplateSupport.md#linked-editing 

In this case it sends several didChange. In master branch you will have one publish diagnostic per changes (if your linked editing range update 3 object parts,y ou will have 3 validation process).

With your current PR you should see just one validation process.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java (hunk 2 above)
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java]: To be honest, I didn't understand the difference between `thenApply` and `thenApplyAsync`. I learned what the difference was, and I changed the code here to use `thenApply`.

Comment (angelozerr) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java]: thenApplyAsync will create a new Thread, I'm not sure that we need that.

**Comment by angelozerr** [comment]
@datho7561 please resolve conflicts.

File: qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java (hunk 1 above)
Comment (datho7561) [reply qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/ls/template/TemplateFileTextDocumentService.java:380]: I still need to make issues to copy it over to lsp4mp, lemminx, ... but I have made the change here

**Comment by angelozerr** [comment]
Many thanks @datho7561 !

**Pull request merged**